Switch to LLVM ORC v2

vtjnash opened the pull request:
Mostly pretty mechanical. Some old hacks went away. Result seems to work, so is presumably good?

vtjnash committed:
LLVM: switch to Orc v2 API

This should not be a user-visible change, but should prepare us for
being able to use new APIs (the legacy ones are deleted now on LLVM
master / v12).

vchuravy commented:
And OrcV1 is gone: http://lists.llvm.org/pipermail/llvm-dev/2020-October/145866.html

vtjnash commented:
remaining tasks are:
 - [ ] switch back to eager finalization (our cgmemmgr doesn't support interleaving of finalization), to address macos64 build failure
 - [ ] figure out symbol conflict from exporting function aliases
 - [ ] figure out why win32 and win64 aren't respecting the `-elf` triple

vchuravy commented:
> figure out why win32 and win64 aren't respecting the -elf triple

In LLVM.jl I had to normalize the triple before passing it to target https://github.com/maleadt/LLVM.jl/blob/b0780fa710219b1ca2e0a418d5dc82fa6dfafe71/src/orc.jl#L130-L134

vtjnash commented:
I suppose, but we don't usually like the default triple, and the process triple that we use appears to already call normalize: https://llvm.org/doxygen/Host_8cpp_source.html#l01545

vtjnash committed:
cmake,LLVM: better xcross configuration

DRY some variables, and pass others along better

vchuravy reviewed (approved):
Looks good to me! Tried this on PPC as well.

Let's squash and GTM?

vchuravy committed:
ExecutionSession no longer takes VModuleKey

vchuravy committed:
Let Orc handle JITEventListerners

vtjnash committed:
llvm: switch from assert to cantFail

Ensures that this usage is proper even if using an LLVM assertions
build, but not using FORCE_ASSERTIONS=1 in Julia

vchuravy commented on src/debuginfo.cpp:
```diff
@@ -316,16 +313,15 @@ class JuliaJITEventListener: public JITEventListener
-        uint64_t SectionLoadOffset = 1; // The real offset shouldn't be 1.
         uint8_t *catchjmp = NULL;
         for (const object::SymbolRef &sym_iter : debugObj.symbols()) {
             StringRef sName;
```
```suggestion
```

vchuravy replied:
```
/cygdrive/c/buildbot/worker/package_win64/build/src/debuginfo.cpp: In member function 'virtual void JuliaJITEventListener::_NotifyObjectEmitted(const llvm::object::ObjectFile&, const llvm::RuntimeDyld::LoadedObjectInfo&, llvm::RTDyldMemoryManager*)':
/cygdrive/c/buildbot/worker/package_win64/build/src/debuginfo.cpp:324:18: error: conflicting declaration 'auto sName'
  324 |             auto sName = cantFail(sym_iter.getName());
      |                  ^~~~~
/cygdrive/c/buildbot/worker/package_win64/build/src/debuginfo.cpp:323:23: note: previous declaration as 'llvm::StringRef sName'
  323 |             StringRef sName;
      |                       ^~~~~
make[1]: *** [Makefile:166: debuginfo.o] Error 1
make[1]: *** Waiting for unfinished jobs....
```

vchuravy commented on src/jitlayers.h:
```diff
@@ -142,58 +139,59 @@ typedef JITSymbol JL_JITSymbol;
 // is expected.
 typedef JITSymbol JL_SymbolInfo;
 
+#if JL_LLVM_VERSION < 12000
```
```suggestion
#if JL_LLVM_VERSION < 120000
```
